### PR TITLE
CURA-8875 Add tie-breaker to seam score calculations

### PR DESCRIPTION
**Pull request by casperlamboo**
to create an even seem in the case where multiple candidate starting positions get the same score.

CURA-8875

### CODE REVIEW

File: src/PathOrderOptimizer.h [hunk 1]
```diff
@@ -501,7 +501,19 @@ class PathOrderOptimizer
                     score += 1000; //1 meter penalty.
                 }
             }
-            if(score < best_score)
+
+            constexpr float EPSILON = 25.0;
+            if(best_score - EPSILON <= score && score <= best_score + EPSILON)
```
Comment (Joeydelarago) [inline src/PathOrderOptimizer.h:506]:
```suggestion
            if(score >= best_score - EPSILON  && score <= best_score + EPSILON)
```
or something like 
`if (fabs(score - best_score) <= EPSILON)`
would be a bit clearer

File: src/PathOrderOptimizer.h [hunk 2]
```diff
@@ -501,7 +501,19 @@ class PathOrderOptimizer
...
+            {
+                // add breaker for two candidate starting location with similar score
+                // if we don't do this then we (can) get an un-even seam
+                Point tie_breaker_point = target_pos != Point(0, 0)
+                    ? Point(0, 0)       // break ties by picking the point that is closest to the origin
+                    : Point(0, 100000); // unless the target position is the origin, then pick some other location
+                best_point = vSize2(best_point - tie_breaker_point) < vSize2(best_point - tie_breaker_point) ? best_point : here;
```
Comment (Joeydelarago) [inline src/PathOrderOptimizer.h:513]: `vSize2(best_point - tie_breaker_point) < vSize2(best_point - tie_breaker_point)`

is always false I think? Did you mean to put something different on one side of this statement?

File: src/PathOrderOptimizer.h [hunk 3]
```diff
@@ -501,7 +501,19 @@ class PathOrderOptimizer
...
+                best_score = std::min(best_score, score);
```
Comment (Joeydelarago) [inline src/PathOrderOptimizer.h:514]: I think you could remove the above code and just leave this in as the tie breaker, that seems to be how it functions now.

**Review by Joeydelarago — CHANGES_REQUESTED** [review]

File: src/PathOrderOptimizer.h (hunk 3 above)
Comment (casperlamboo) [reply src/PathOrderOptimizer.h:514]: Well there can still be a small difference (at most epsilon) between the two numbers, I'd prefer to keep it.

File: src/PathOrderOptimizer.h (hunk 2 above)
Comment (casperlamboo) [reply src/PathOrderOptimizer.h:513]: oops, should be
```suggestion
                 best_point = vSize2(best_point - tie_breaker_point) < vSize2(here - tie_breaker_point) ? best_point : here; 
```

Comment (casperlamboo) [reply src/PathOrderOptimizer.h:513]: https://github.com/Ultimaker/CuraEngine/commit/4e137f5273716dedea5678f32183a601bc71fb6e

File: src/PathOrderOptimizer.h (hunk 1 above)
Comment (Joeydelarago) [reply src/PathOrderOptimizer.h:506]: I see where you are coming from, it is a lot easier to read when I think about it in that format 👌

Comment (casperlamboo) [reply src/PathOrderOptimizer.h]: https://github.com/Ultimaker/CuraEngine/pull/1623/commits/fcf4eac5c4177f0d47b43383bf8e91db7b4cad70

**Review by Joeydelarago — APPROVED** [review]

**Pull request merged**